library/Fission/Types.hs

```diff
@@ -396,33 +399,48 @@ instance User.Creator Fission where
           Left err -> Error.relaxedLeft err
           Right _  -> Right userId
 
+runUpdate ::
+     Fission (Either User.Modifier.Errors a)
+  -> UserId
+  -> Text
+  -> Text
+  -> Fission (Either User.Modifier.Errors a)
+runUpdate updateDB uID subdomain txt =
+  updateDB >>= \case
+    Left err ->
+      return $ Left err
+  
+    Right dbVal -> do
+      runDB (User.getById uID) >>= \case
+        Nothing ->
+          return . Error.openLeft $ NotFound @User
+
+        Just (Entity _ User { userUsername = Username rawUN }) -> do
+          domainName <- asks userRootDomain
+          zoneID     <- asks userZoneID
+
+          let
+            url = URL {domainName, subdomain = Just (Subdomain subdomain) <> Just (Subdomain rawUN)}
+            segments = DNS.splitRecord txt
+
+          Route53.set Txt url zoneID segments 10 <&> \case
+            Left serverErr -> Error.openLeft serverErr
+            Right _        -> Right dbVal
+
+
 instance User.Modifier Fission where
   updatePassword uID pass now =
     runDB $ User.updatePassword uID pass now
 
   updatePublicKey uID pk now =
-    runDB (User.updatePublicKey uID pk now) >>= \case
-      Left err ->
-        return $ Left err
+    runUpdate (User.updatePublicKey uID pk now) uID "_did" keyTxt 
+    where 
+      keyTxt = textDisplay $ DID Key pk
 
-      Right _ -> do
-        runDB (User.getById uID) >>= \case
-          Nothing -> 
-            return . Error.openLeft $ NotFound @User
-
-          Just (Entity _ User { userUsername = Username rawUN }) -> do
-            domainName <- asks userRootDomain
-            zoneID     <- asks userZoneID
-
-            let
-              subdomain = Just $ Subdomain rawUN
-              url       = URL {domainName, subdomain = Just (Subdomain "_did") <> subdomain}
-              did       = textDisplay (DID pk Key)
-              didSegments = DNS.splitRecord did
-
-            Route53.set Txt url zoneID didSegments 10 <&> \case
-              Left serverErr -> Error.openLeft serverErr
-              Right _        -> Right pk
+  updateExchangeKeys uID keys now =
+    runUpdate (User.updateExchangeKeys uID keys now) uID "_exchange" keysTxt
+    where
+      keysTxt = Text.intercalate "," $ textDisplay . DID Key . Key.RSAPublicKey <$> keys
```

Comment:
Yup, this works! I’ve found elsewhere that`x $ y . z <$> a` can be difficult to parse. Consider parens:

```haskell
updatePublicKey uID pk now = runUpdate updatePK uID "_exchange" keyTxt
  where
    updatePK = User.updatePublicKey uID pk now
    keysTxt = Text.intercalate “,” (textDisplay . DID Key . Key.RSAPublicKey <$> keys)
```